python/pyarrow/_compute.pyx

```diff
@@ -2182,3 +2255,175 @@ cdef CExpression _bind(Expression filter, Schema schema) except *:
+        types for the function
+    out_type : DataType
+        output type of the function
+    callback : callable
+        user defined function
+        function includes arguments equal to the number
+        of input_types defined. The return type of the 
+        function is of the type defined as output_type. 
+        The output is a datum object which can be
+        an Array or a ChunkedArray or a Table or a RecordBatch.
+
+    Example
+    -------
+
+    >>> from pyarrow import compute as pc
+    >>> from pyarrow.compute import register_function
+    >>> from pyarrow.compute import InputType
+    >>> 
+    >>> func_doc = {}
+    >>> func_doc["summary"] = "simple udf"
+    >>> func_doc["description"] = "add a constant to a scalar"
+    >>> func_doc["arg_names"] = ["x"]
+    >>> 
+    >>> def add_constant(array):
+    ...     return pc.call_function("add", [array, 1])
+    ... 
+    >>> 
+    >>> func_name = "py_add_func"
+    >>> arity = 1
+    >>> in_types = [InputType.array(pa.int64())]
+    >>> out_type = pa.int64()
+    >>> register_function(func_name, arity, func_doc,
+    ...                   in_types, out_type, add_constant)
+    >>> 
+    >>> func = pc.get_function(func_name)
+    >>> func.name
+    'py_add_func'
+    >>> ans = pc.call_function(func_name, [pa.array([20])])
+    >>> ans
+    <pyarrow.lib.Int64Array object at 0x10c22e700>
+    [
+    21
+    ]
+    """
+    cdef:
+        c_string c_func_name
+        CArity c_arity
+        CFunctionDoc c_func_doc
+        CInputType in_tmp
+        vector[CInputType] c_in_types
+        PyObject* c_callback
+        shared_ptr[CDataType] c_type
+        COutputType* c_out_type
+        CScalarUdfBuilder* c_sc_builder
+        CStatus st
+        CScalarUdfOptions* c_options
+        object obj
+
+    if func_name and isinstance(func_name, str):
+        c_func_name = tobytes(func_name)
+    else:
+        raise ValueError("func_name should be str")
+
+    if num_args and isinstance(num_args, int):
+        assert num_args > 0
+        if num_args == 0:
+            c_arity = CArity.Nullary()
+        elif num_args == 1:
+            c_arity = CArity.Unary()
+        elif num_args == 2:
+            c_arity = CArity.Binary()
+        elif num_args == 3:
+            c_arity = CArity.Ternary()
+        elif num_args > 3:
+            c_arity = CArity.VarArgs(num_args)
+    else:
+        raise ValueError("arity must be an instance of Arity")
+
+    c_func_doc = _make_function_doc(function_doc)
+
+    if in_types and isinstance(in_types, list):
+        for in_type in in_types:
+            in_tmp = (<InputType> in_type).input_type
+            c_in_types.push_back(in_tmp)
+    else:
+        raise ValueError("input types must be of type InputType")
+
+    if out_type:
+        c_type = pyarrow_unwrap_data_type(out_type)
+    else:
+        raise ValueError("Output value type must be defined")
+
+    if callback and callable(callback):
```

Comment:
```suggestion
    if callable(callback):
```

Reply:
The intent here is that if callback is the wrong type or None, `callable` will still catch it. There's no need to separately check for none.